hamilton/function_modifiers/macros.py

```diff
@@ -773,65 +809,152 @@ def final_result(upstream_int: int) -> int:
     ) -> Tuple[List[node.Node], Dict[str, str]]:
         """Injects nodes into the graph. This creates a node for each pipe() step,
         then reassigns the inputs to pass it in."""
-        sig = inspect.signature(fn)
-        first_parameter = list(sig.parameters.values())[0].name
-        # use the name of the parameter to determine the first node
-        # Then wire them all through in order
-        # if it resolves, great
-        # if not, skip that, pointing to the previous
-        # Create a node along the way
-        if first_parameter not in params:
-            raise base.InvalidDecoratorException(
-                f"Function: {fn.__name__} has a first parameter that is not a dependency. "
-                f"@pipe requires the parameter names to match the function parameters. "
-                f"Thus it might not be compatible with some other decorators"
-            )
+        parameters_transforms_map = self._filter_individual_target(params=params)
+
+        # This reverts back to legacy chaining through first parameter
+        if not parameters_transforms_map:
+            sig = inspect.signature(fn)
+            first_parameter = list(sig.parameters.values())[0].name
+            # use the name of the parameter to determine the first node
+            # Then wire them all through in order
+            # if it resolves, great
+            # if not, skip that, pointing to the previous
+            # Create a node along the way
+            if first_parameter not in params:
+                raise base.InvalidDecoratorException(
+                    f"Function: {fn.__name__} has a first parameter that is not a dependency. "
+                    f"@pipe requires the parameter names to match the function parameters. "
+                    f"Thus it might not be compatible with some other decorators"
+                )
+            parameters_transforms_map[first_parameter] = self.transforms
+        else:  # checks that each transform has at least one target parameter
+            transform_set = []
+            for param in parameters_transforms_map:
+                transform_set.extend(parameters_transforms_map[param])
+            transform_set = set(transform_set)
+            if len(transform_set) != len(self.transforms):
+                raise MissingTargetError(
+                    "The on_input settings are unclear. Please make sure all transforms "
+                    "either have specified individually or globally a target or there is "
+                    "no on_input usage."
+                )
 
-        # Chaining gets done by linking the first argument of each node
-        nodes, current_param = chain_transforms(
-            first_arg=first_parameter,
-            transforms=self.transforms,
-            namespace=self.namespace,
-            config=config,
-            fn=fn,
-        )
+        total_nodes = []
+        total_rename_maps = {}
+
+        for param in parameters_transforms_map:
```

Comment:
This logic is getting complex -- we should probably break it out into a separate function and add a unit test or two for it...